### ISSUE TITLE
Testing framework problems

### ISSUE THREAD

**Issue by davidmanzanares** [CartoDB/renderer-prototype#151]
Current issues/problems:

Moved to https://github.com/CartoDB/renderer-prototype/issues/156:
- [x] Lack of GeoJSON properties. It is impossible or painful to test many styling expressions without these.
- [x] Lack of GeoJSON metadata. It is impossible to reliable test many styling expressions without this, particularly categories.

Moved to https://github.com/CartoDB/renderer-prototype/issues/158:
- [x] An easy way to run and/or prepare just one test
- [x] A way to load multiple GeoJSONs from the same test. The current approach only allows for one GeoJSON, that is not enough for testing multilayer.

Moved to https://github.com/CartoDB/renderer-prototype/pull/161:
- [x] Display image difference (not more `true is not false` messages).
- [x] Log Chrome's console output

cc: @Jesus89 @IagoLast @rochoa

**Comment by Jesus89** [comment]
Proposals:
- Using `fscenario.js` for the test you want to filter?
- Add a directory `sources` with different geojson files that can be used across the render tests
- Add `diff.png` image for each test

I agree with the rest of missing parts.

**Comment by davidmanzanares** [comment]
> Using fscenario.js for the test you want to filter?

What do you mean?


> Add a directory sources with different geojson files that can be used across the render tests

Agree 100%

> Add diff.png image for each test

Agree, but I'd still want a nice number on the console error log.

**Comment by Jesus89** [comment]
>> Using fscenario.js for the test you want to filter?

> What do you mean?

For the tests detection all the `scenario.js` files are used. So in order to run one specific tests it could be marked as `fscenario.js` (the file name).

**Comment by IagoLast** [comment]
@Jesus89  can we find a way to integrate our framework in a standard test framework like jasmine/mocha/jest instead reimplementing it?

**Comment by IagoLast** [comment]
I like the idea of a semantic folder name : 

`x-<scenario-folder-name>` : skip the test
`f-<scenario-folder-name>` : skip tests whose folder name does not start with `f-`

**Comment by IagoLast** [comment]
Also I dont se the need for  having the `diff.png` inegrated in the framework.

- Is a lot of work to implement this feature
- the Normal scenario is to have all tests passing and an error should be the exception. If a tests fails you have the `reference.png` and the `_out.png` files and with imagemagik we can easily generate a  `diff.png`

**Comment by Jesus89** [comment]
@IagoLast we are already using mocha for the render tests but if you have proposals to use other frameworks go ahead :)

We can use semantic folder but the problem is that the folder is the name of the tests so it is more focused with the file name.

If we have seen the need to have the diff.png image we should implement it. Btw, this should be done in the Exquisite project and also the output returning how many pixels (or %) are failing.

**Comment by IagoLast** [comment]
> We can use semantic folder but the problem is that the folder is the name of the tests so it is more focused with the file name.

IMO folders should work as `describe` do in the most of test frameworks allowing multiple levels.

I know is a lot of work, so IMO we can provide this funcionality only at the lowest level.

PD: stop calling files to what are really folders! ;) 

```
├── render
│   ├── f-line-string   # Only run this
│   │   ├── color
│   │   │   ├── reference.png
│   │   │   └── scenario.js
│   │   ├── data.geojson
│   │   ├── default
│   │   │   ├── reference.png
│   │   │   └── scenario.js
│   │   └── x-width    # skip this
│   │       ├── reference.png
│   │       └── scenario.js
│   ├── multi-line-string

```

> If we have seen the need to have the diff.png image we should implement it. Btw, this should be done in the Exquisite project and also the output returning how many pixels (or %) are failing.

https://github.com/CartoDB/renderer-prototype/issues/151#issuecomment-372643285

**Comment by Jesus89** [comment]
I'm going to implement the `x-folder` `f-folder` detection, hope we are all agree on that :)

Btw, all the things related to the chrome's console output, the diff image detection and rich output information should be done in Exquisite if we finally want to do it.

**Comment by davidmanzanares** [comment]
I'll work on GeoJSON metadata and GeoJSON properties then

**Comment by Jesus89** [comment]
Closed by #156, #158, #161
